engines/scumm/imuse_digi/dimuse_internalmixer.cpp

```diff
@@ -0,0 +1,1093 @@
+* as published by the Free Software Foundation; either version 2
+* of the License, or (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program; if not, write to the Free Software
+* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*
+*/
+
+#include "common/scummsys.h"
+#include "common/mutex.h"
+#include "common/serializer.h"
+#include "common/textconsole.h"
+#include "common/util.h"
+
+#include "scumm/imuse_digi/dimuse_engine.h"
+#include "scumm/imuse_digi/dimuse_internalmixer.h"
+
+namespace Scumm {
+
+IMuseDigiInternalMixer::IMuseDigiInternalMixer(Audio::Mixer *mixer) {
+	_stream = Audio::makeQueuingAudioStream(22050, true);
+	_mixer = mixer;
+	_radioChatter = 0;
+	_amp8Table = NULL;
+}
+
+IMuseDigiInternalMixer::~IMuseDigiInternalMixer() {
+	free(_amp8Table);
+	_amp8Table = NULL;
+}
+
+// Lookup table for a linear volume ramp (0 to 16) accounting for panning (-8 to 8)
+static const int8 _stereoVolumeTable[284] = {
+	0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
+	0,  0,  0,  1,  1,  1,  1,  1,  1,  2,  2,  2,  2,  2,  2,  2,  2,  0,  0,  1,  1,  1,  1,  2,  2,  2,
+	2,  2,  3,  3,  3,  3,  3,  3,  0,  0,  1,  1,  2,  2,  2,  3,  3,  3,  3,  4,  4,  4,  4,  4,  4,  0,
+	0,  1,  1,  2,  2,  3,  3,  4,  4,  4,  4,  5,  5,  5,  5,  5,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,
+	5,  5,  6,  6,  6,  6,  6,  0,  1,  1,  2,  3,  3,  4,  4,  5,  5,  6,  6,  6,  7,  7,  7,  7,  0,  1,
+	2,  2,  3,  4,  4,  5,  6,  6,  7,  7,  7,  8,  8,  8,  8,  0,  1,  2,  3,  3,  4,  5,  6,  6,  7,  7,
+	8,  8,  9,  9,  9,  9,  0,  1,  2,  3,  4,  5,  6,  6,  7,  8,  8,  9,  9,  10, 10, 10, 10, 0,  1,  2,
+	3,  4,  5,  6,  7,  8,  9,  9,  10, 10, 11, 11, 11, 11, 0,  1,  2,  3,  5,  6,  7,  8,  8,  9,  10, 11,
+	11, 11, 12, 12, 12, 0,  1,  3,  4,  5,  6,  7,  8,  9,  10, 11, 11, 12, 12, 13, 13, 13, 0,  1,  3,  4,
+	5,  7,  8,  9,  10, 11, 12, 12, 13, 13, 14, 14, 14, 0,  1,  3,  4,  6,  7,  8,  10, 11, 12, 12, 13, 14,
+	14, 15, 15, 15, 0,  2,  3,  5,  6,  8,  9,  10, 11, 12, 13, 14, 15, 15, 16, 16, 16, 0,  0,  0
+};
+
+int IMuseDigiInternalMixer::init(int bytesPerSample, int numChannels, uint8 *mixBuf, int mixBufSize, int sizeSampleKB, int mixChannelsNum) {
+	int amplitudeValue;
+	int waveMixChannelsCount;
+	int softLdenominator;
+	int softLnumerator;
+	int softLcurValue;
+	int zeroCenterOffset;
+
+	_outWordSize = bytesPerSample;
+	_outChannelCount = numChannels;
+	_mixBufSize = mixBufSize;
+	_stereoReverseFlag = sizeSampleKB;
+	_mixBuf = mixBuf;
+
+	waveMixChannelsCount = mixChannelsNum;
+
+	int *tableMalloc = (int *)malloc(213504);
```

Comment:
are you able to add some comments about numbers?

Reply:
Done